### PR TITLE
Add metadata to literal

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
NA

## Why are the changes needed?
Add metadata attribute to `Literal` class, so users can add arbitrary information to the literal.

## What changes were proposed in this pull request?
Update literal class

## How was this patch tested?
Unit tests

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
NA

## Docs link
NA

### CODE REVIEW

File: flytekit/models/literals.py [hunk 1]
```diff
@@ -938,4 +953,5 @@ def from_flyte_idl(cls, pb2_object):
             collection=collection,
             map=LiteralMap.from_flyte_idl(pb2_object.map) if pb2_object.HasField("map") else None,
             hash=pb2_object.hash if pb2_object.hash else None,
+            metadata={k: v for k, v in pb2_object.metadata.items()},
```
Comment (wild-endeavor) [inline flytekit/models/literals.py:956]: can we make this None if it doesn't exist in the proto?

**Review by wild-endeavor — APPROVED** [review]

**Comment by cameronraysmith** [comment]
@pingsutw I'll look further later and file an issue if needed, but I bisected

```pytb
AttributeError: 'Literal' object has no attribute '_metadata'
```

to this PR when executing locally with the cache enabled. 
The error does not occur in 1.10.3, does occur up through 1.10.8b4, resolves when disabling the cache, and I can patch it in pdb with `input_literal_map.literals['task_output_name']._metadata = None` before https://github.com/flyteorg/flytekit/blob/64b8468fa3ed89b7647936e80dd30e6c0ad7c6ed/flytekit/core/base_task.py#L282 because the error occurs when attempting to generate the string representation of `input_literals_map` there.

Maybe the metadata is not being set to `None` by default when loading from the cache?
I tried installing from master and the error persists.

<details><summary>Details</summary>
<p>

╭────── Traceback (most recent call last) ────────╮
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/exceptions/scopes.py:219 in user_entry_point                     │
│                                                                                                                                                                        │
│ ❱ 219 │   │   │   │   return wrapped(*args, **kwargs)                                                                                                                  │
│                                                                                                                                                                        │
│ /nix/store/awfn4iq4yqr6gpc6yadd36wgxyz3dsbh-source/src/pyrovelocity/workflows/main_workflow.py:210 in module_workflow                                                  │
│                                                                                                                                                                        │
│ ❱ 210 │   processed_data = preprocess_data(                                                                                                                            │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/core/base_task.py:315 in __call__                                │
│                                                                                                                                                                        │
│ ❱ 315 │   │   return flyte_entity_call_handler(self, *args, **kwargs)  # type: ignore                                                                                  │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/core/promise.py:1191 in flyte_entity_call_handler                │
│                                                                                                                                                                        │
│ ❱ 1191 │   │   │   return cast(LocallyExecutable, entity).local_execute(ctx, **kwargs)                                                                                 │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/core/base_task.py:272 in local_execute                           │
│                                                                                                                                                                        │
│ ❱ 272 │   │   │   │   f"Checking cache for task named {self.name}, cache version {self.metadat                                                                         │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/models/common.py:54 in __str__                                   │
│                                                                                                                                                                        │
│ ❱  54 │   │   return self.verbose_string()                                                                                                                             │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/models/common.py:70 in verbose_string                            │
│                                                                                                                                                                        │
│ ❱  70 │   │   return self.short_string()                                                                                                                               │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/models/common.py:63 in short_string                              │
│                                                                                                                                                                        │
│ ❱  63 │   │   literal_str = re.sub(r"\s+", " ", str(self.to_flyte_idl())).strip()                                                                                      │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/models/literals.py:692 in to_flyte_idl                           │
│                                                                                                                                                                        │
│ ❱ 692 │   │   return _literals_pb2.LiteralMap(literals={k: v.to_flyte_idl() for k, v in self.l                                                                         │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/models/literals.py:692 in <dictcomp>                             │
│                                                                                                                                                                        │
│ ❱ 692 │   │   return _literals_pb2.LiteralMap(literals={k: v.to_flyte_idl() for k, v in self.l                                                                         │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/models/literals.py:937 in to_flyte_idl                           │
│                                                                                                                                                                        │
│ ❱ 937 │   │   │   metadata=self.metadata,                                                                                                                              │
│                                                                                                                                                                        │
│ /nix/store/1ragwbwz170v6vpajkc1bgh2mswdwi5a-python3-3.10.13-env/lib/python3.10/site-packages/flytekit/models/literals.py:926 in metadata                               │
│                                                                                                                                                                        │
│ ❱ 926 │   │   return self._metadata                                                                                                                                    │
╰─────────────╯
AttributeError: 'Literal' object has no attribute '_metadata'

</p>
</details>

File: flytekit/models/literals.py [hunk 2]
```diff
@@ -912,6 +918,14 @@ def hash(self):
     def hash(self, value):
         self._hash = value
 
+    @property
+    def metadata(self):
+        """
+        This value holds metadata about the literal.
+        :rtype: typing.Dict[str, str]
+        """
+        return self._metadata
```
Comment (cameronraysmith) [inline flytekit/models/literals.py:927]:
```suggestion
return getattr(self, '_metadata', None)
```

fixes https://github.com/flyteorg/flytekit/pull/2147#issuecomment-1985257645

```pytb
AttributeError: 'Literal' object has no attribute '_metadata'
```

but I would imagine there is a better place to enforce initialization

**Comment by cameronraysmith** [comment]
Upgrading 1.10.3 to 1.10.7 requires deleting/refreshing the local cache db `~/.flyte/local-cache/cache.db` or this change can lead to

```pytb
AttributeError: 'Literal' object has no attribute '_metadata'
```

**Pull request merged**